### DISCUSSION TITLE
Over fetching async atoms w jotai-recoil

### DISCUSSION THREAD

**Discussion by stevemolitor** [pmndrs/jotai#2164] [General] [answered]
I have an async atom that is mounting and unmounting. This causes HTTP fetch requests the atom to fire too often.

In the atom docs it says:

> In some edge cases, an atom can be unmounted and then mounted immediately.

What are those edge cases?

Unfortunately I haven't been able to pull my case out into a sandbox I can share yet. I'm hoping to understand in a general way when module level atoms would unmount.

**Comment by dai-shi** [comment]
> What are those edge cases?

I think it's when atom dependencies change dynamically.

> This causes HTTP fetch requests the atom to fire too often.

That sounds bad. I think there should be a way to work around (assuming if it's not a bug.)

> when module level atoms would unmount.

If an atom isn't depended by any other mounted atoms, it will unmount.

**Comment by stevemolitor** [comment]
Hmm thanks, maybe I'm misunderstanding the implications of unmounting.

Here is a [CodeSandbox](https://codesandbox.io/s/great-tristan-ll6l72?file=/src/App.tsx). I conditionally render a component that uses an async atom:

```ts
const DataOrStatic = ({ showData }: { showData: boolean }) => {
  if (showData) {
    return <ShowData />;
  }
  return <ShowStaticValue />;
};
```

```ts
const ShowData = () => {
  const data = useAtomValue(dataAtom);
  return <div>data: {data}</div>;
};
```

I log the mounting and unmounting of the async atom:

```ts
const fetchData = async () => {
  return new Promise<string>((resolve) => {
    setTimeout(() => {
      console.log("loading data…");
      return resolve("data-here");
    }, 1000);
  });
};

const dataAtom = atomWithDefault(fetchData);

dataAtom.onMount = () => {
  console.log("mounting data atom");
  return () => {
    console.log("unmounting data atom");
  };
};
```

I see the mounting, unmount log statements as I toggle. But, it does _not_ refetch in this CodeSandbox:

![](https://cdn.zappy.app/490cc39a8f7d01f65dc36ef7072f36e9.png)

![](https://cdn.zappy.app/51f508634f817fabd04f78745982221d.png)

Why not? How does it know that it doesn't need to re-fetch, even though the atom has unmounted and remounted? I'm starting to think the issue in my app isn't unmounting, but maybe something else.

Thanks again!

**Reply by dai-shi** [reply]
It will unmount, but it doesn't throw away the value. Yeah, I don't think you need to care about mounting in your case.

**Comment by stevemolitor** [comment]
I should add, my app is using [jotai-recoil](https://github.com/jotaijs/jotai-recoil) to gradually migrate from Recoil to Jotai. The issue with refetching the async `atomWithDefault` atom happens when it has a dependency (calls `get`) on an atom created via `atomViaRecoilValue`. 

I'll try harder to recreate a failing example outside of my (very large) app.

**Reply by dai-shi** [reply]
Oh, it's very nice to hear that you use jotai-recoil. I developed it without knowing if it's really helpful.

**Comment by stevemolitor** [comment]
I've reproduced the issue, using `jotai-recoil` with [this CodeSandbox](https://codesandbox.io/s/flamboyant-sanderson-kp8k3q?file=/src/App.tsx).

Re-fetching of an async Jotai `atomWithDefault` atom happens when:

1. I unmount the component using the Jotai atom.
2. I change _unrelated_ Recoil state.
3. I remount the component using the Jotai atom. This cause the Jotai atom to re-fetch / re-call its initialization function.

The offending line is the `get` call here:

```ts
const dataAtom = atomWithDefault(async (get) => {
  get(jotaiNeverChangesAtom);
  return fetchData();
});
```

The `jotaiNeverChanges` atom use `atomWithRecoil` value to a Recoil atom that never changes. However changing _another_ Recoil atom causes the effect in `useSyncRecoilSnapshot` to set a new snapshot (as expected), which seems to cause `dataAtom` to re-evaluate its read function (not expected).

Very strange!

![](https://cdn.zappy.app/4a7c8b4fc3f563b056dd951e169965bb.gif)

**Reply by dai-shi** [reply]
So, it's specifically an issue with `jotai-recoil`. Can you change the discussion title please?

**Reply by stevemolitor** [reply]
Sorry about the misleading title. I changed the title.

**Reply by dai-shi** [reply, marked as answer]
https://codesandbox.io/s/pensive-mcclintock-v65j88 works with Jotai v2.2.3
This seems like a bug since Jotai v2.3.0. 😢

**Reply by dai-shi** [reply]
Opened a new issue #2168.

**Reply by stevemolitor** [reply]
Thanks for the quick investigation!

**Reply by dai-shi** [reply]
#2197 will fix your issue: https://codesandbox.io/s/heuristic-lamarr-gpqylx
